File: backend/app/rbac/permissions.py
"""Action -> allowed roles map.  Mirrors frontend src/rbac/permissions.js."""
from app.rbac.roles import Role

PERMISSIONS: dict[str, list[Role]] = {
    "create_draft":               [Role.EXECUTIVE],
    "save_draft_details":         [Role.EXECUTIVE],
    "submit_details_for_review":  [Role.EXECUTIVE],
    "upload_loi":                 [Role.EXECUTIVE],
    "view_own_loi":               [Role.EXECUTIVE],

    "shortlist":                  [Role.SUPERVISOR, Role.SUB_SUPERVISOR],
    "approve_details":            [Role.SUPERVISOR, Role.SUB_SUPERVISOR],
    "reject":                     [Role.SUPERVISOR, Role.SUB_SUPERVISOR],
    "archive":                    [Role.SUPERVISOR, Role.SUB_SUPERVISOR],
    "set_loi_timeline":           [Role.SUPERVISOR, Role.SUB_SUPERVISOR],
    "push_to_payments":           [Role.SUPERVISOR, Role.SUB_SUPERVISOR],
    "reassign_site":              [Role.SUPERVISOR],
    "assign_sub_supervisor":      [Role.SUPERVISOR],
}


def can(role: Role, action: str) -> bool:
    allowed = PERMISSIONS.get(action, [])
    return role in allowed

File: frontend/src/rbac/permissions.js
import { ROLE } from './roles.js';

// ACTION -> allowed roles map
// Keep in sync with backend app/rbac/permissions.py
export const PERMISSIONS = {
  create_draft:               [ROLE.EXECUTIVE],
  save_draft_details:         [ROLE.EXECUTIVE],
  submit_details_for_review:  [ROLE.EXECUTIVE],
  upload_loi:                 [ROLE.EXECUTIVE],
  view_own_loi:               [ROLE.EXECUTIVE],

  shortlist:                  [ROLE.SUPERVISOR, ROLE.SUB_SUPERVISOR],
  approve_details:            [ROLE.SUPERVISOR, ROLE.SUB_SUPERVISOR],
  reject:                     [ROLE.SUPERVISOR, ROLE.SUB_SUPERVISOR],
  archive:                    [ROLE.SUPERVISOR, ROLE.SUB_SUPERVISOR],
  set_loi_timeline:           [ROLE.SUPERVISOR, ROLE.SUB_SUPERVISOR],
  push_to_payments:           [ROLE.SUPERVISOR, ROLE.SUB_SUPERVISOR],
  reassign_site:              [ROLE.SUPERVISOR],
  assign_sub_supervisor:      [ROLE.SUPERVISOR],
};

export function can(role, action) {
  const allowed = PERMISSIONS[action];
  if (!allowed) return false;
  return allowed.includes(role);
}
